### DISCUSSION TITLE
Mixed calling of nonblocking SSL_read/SSL_write

### DISCUSSION THREAD

**Discussion by SR42** [openssl/openssl#23112] [Q&A] [answered]
Hi! Current documentation says this about safety of mixed calling nonblocking SSL_read() and SSL_write():

> It is safe to call SSL_read() or SSL_read_ex() when more data is available even when the call that set this error was an SSL_write() or SSL_write_ex(). However, if the call was an SSL_write() or SSL_write_ex(), it should be called again to continue sending the application data. If you get SSL_ERROR_WANT_WRITE from SSL_write() or SSL_write_ex() then you should not do any other operation that could trigger IO other than to repeat the previous SSL_write() call.

Another words, it can be presented in the following format:
First call -> Error -> Second call : Safety

SSL_write() -> SSL_ERROR_WANT_READ -> SSL_read() : SAFE 🟢
SSL_write() -> SSL_ERROR_WANT_WRITE -> SSL_read() : UNSAFE 🔴

However, I could not deduce other scenarios from the documentation:
SSL_read() -> SSL_ERROR_WANT_READ -> SSL_write() : ???
SSL_read() -> SSL_ERROR_WANT_WRITE -> SSL_write() : ???

Could you fill this gap?

**Comment by mattcaswell** [comment, marked as answer]
Yeah - reading it now that documentation is quite confusing. The first two sentences that you quote seem to contradict each other.

The problematic scenario is getting SSL_ERROR_WANT_WRITE from SSL_write(). In that case you must only ever call SSL_write() afterwards (i.e. never SSL_read). All the other scenarios you list are safe.

**Comment by markl-arista** [comment]
@mattcaswell 

Hi Matt. Really appreciate your answer. I agree with your assessment that the first two sentences in the documentation pointed out by @SR42 seem to contradict each other.

In addition, according to doc for [`SSL_write`](https://docs.openssl.org/3.0/man3/SSL_write) and doc for [`SSL_read`](https://docs.openssl.org/3.0/man3/SSL_read/#notes), one may get the impression that a `SSL_read` (or `SSL_write`) call that returns either `SSL_ERROR_WANT_READ` or `SSL_ERROR_WANT_WRITE` must be repeated again. However, according to your answer, 

> SSL_write() -> SSL_ERROR_WANT_READ -> SSL_read() : SAFE 🟢

should be fine. That is, there is no requirement that the next action must be a retry of `SSL_write`. 

So... am I right that the **ONLY** thing one **MUST** do is that when `SSL_write` returns `SSL_ERROR_WANT_WRITE`, the next call **MUST** be `SSL_write` and **MUST** be with the same argument (according to https://github.com/openssl/openssl/discussions/25310).
